lib/fluent/plugin/auditd.rb

```diff
@@ -0,0 +1,123 @@
+module Fluent
+  class Auditd
+    class AuditdParserException < StandardError
+    end
+    
+    MSG = "msg"
+    TIMESTAMP = "time"
+    AUDIT = "audit"
+
+    def parse_auditd_line(line)
+      states = {}
+      states[:start] = 0
+      states[:key] = 1
+      states[:value] = 2
+      state = states[:start]
+      stack = ["$"]
+      line << "$"
+      nested = false
+      key, nested_key, value = nil
+      result = {}
+      for i in 0...line.length do
```

Comment:
I realize you are trying to create a generic parser but might it not be easier to user regex to identify the specific audit entry we desire and to further pull out and parse the values we require?  With use of regex and string.split it seems we could significantly get over needing to process the line byte by byte

I might suggest if we continue processing in this fashion to break each section we are processing into its own method to make the entire code more digestable